File: Algorithm/Sort/CyclicSort/leetcode/41. First Missing Positive/GPT/First-Missing-Positive.py
# 以下は **Python (CPython 3.11.4)** / **LeetCode 環境** / **Pylance 型チェック対応** の
# `O(n)` 時間・`O(1)` 追加メモリの解法です。

# ---

# ```python
from typing import List

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        """
        配列 nums に存在しない最小の正の整数を返す

        :param nums: List[int] - 整数配列（順不同、負数や0を含む可能性あり）
        :return: int - 配列に存在しない最小の正の整数

        時間計算量: O(n)
        空間計算量: O(1) 追加メモリ（インプレース処理）
        対応環境: Python (CPython 3.11.4) / LeetCode
        """
        n: int = len(nums)

        # 値を正しい位置に移動する（Cyclic Sort）
        for i in range(n):
            while (
                1 <= nums[i] <= n and
                nums[nums[i] - 1] != nums[i]
            ):
                correct_index: int = nums[i] - 1
                nums[i], nums[correct_index] = nums[correct_index], nums[i]

        # 最初に位置と値が一致しないインデックス+1が答え
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1

        # すべて一致すれば n+1
        return n + 1

# ## 実装ポイント

# 1. **Cyclic Sort（配置スワップ法）**

#    * `1 <= x <= n` の値だけを正しい位置（`x-1` インデックス）に置く
#    * スワップ後も条件を満たせば繰り返す
# 2. **O(n) 時間**

#    * 各要素は最大1回だけ正しい位置に移動
# 3. **O(1) 追加メモリ**

#    * 配列内で直接スワップするため、追加配列は不要
# 4. **Pylance 型安全**

#    * 引数・変数に型を明示 (`List[int]`, `int`)

# ---

# もし希望であれば、この Python 版についても **図解トレース** を作って
# どのように要素が移動していくかを視覚的に説明できます。

File: Algorithm/Sort/CyclicSort/leetcode/41. First Missing Positive/GPT/First-Missing-Positive.ts
// 以下は **TypeScript 5.1 / Node.js 18.16.1** 対応、**O(n) 時間・O(1) 追加メモリ** の解法です。
// 配列をインプレースで並べ替える **Cyclic Sort（配置スワップ法）** を利用しています。

// ```typescript
/**
 * 配列 nums に存在しない最小の正の整数を返す
 * @param nums - 整数配列（順不同、負数や0を含む可能性あり）
 * @returns 配列に存在しない最小の正の整数
 *
 * 時間計算量: O(n)
 * 空間計算量: O(1) 追加メモリ
 * 対応環境: TypeScript 5.1 / Node.js 18.16.1 / LeetCode
 */
function firstMissingPositiveGPT(nums: number[]): number {
    const n: number = nums.length;

    // 正しい位置に [1, n] の範囲の値を移動
    for (let i = 0; i < n; i++) {
        while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] !== nums[i]) {
            const correctIndex: number = nums[i] - 1;
            [nums[i], nums[correctIndex]] = [nums[correctIndex], nums[i]];
        }
    }

    // 値とインデックス+1が一致しない最初の位置が答え
    for (let i = 0; i < n; i++) {
        if (nums[i] !== i + 1) {
            return i + 1;
        }
    }

    // 全て揃っていれば n+1
    return n + 1;
}

// ### アルゴリズム概要

// 1. **配置スワップ法**

//    * 値 `x` が `1 <= x <= n` であれば、インデックス `x-1` に置く。
//    * スワップ後も同じ条件に該当すれば再度スワップ（`while`ループ）。
// 2. **O(n) 時間**

//    * 各要素は最大1回しか正しい位置に移動しない。
// 3. **O(1) 追加メモリ**

//    * 配列を直接並べ替えるため、追加の配列は不要。
// 4. **範囲外や重複は無視**

//    * `[1, n]` の範囲外は位置合わせ対象外。
